// stream-chat-react/dist/Audio.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Audio = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Audio = function (_React$Component) {
  (0, _inherits2.default)(Audio, _React$Component);

  var _super = _createSuper(Audio);

  function Audio(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Audio);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "playAudio", function () {
      if (_this.audioRef.current !== null) {
        _this.audioRef.current.pause();

        _this.updateProgress();

        _this.setState({
          playing: true,
          updateProgress: setInterval(_this.updateProgress, 500)
        });

        _this.audioRef.current.play();
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "pauseAudio", function () {
      if (_this.audioRef.current !== null) {
        _this.audioRef.current.pause();
      }

      _this.setState({
        playing: false
      });

      window.clearInterval(_this.state.updateProgress);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "updateProgress", function () {
      if (_this.audioRef.current !== null) {
        var position = _this.audioRef.current.currentTime;
        var duration = _this.audioRef.current.duration;
        var progress = 100 / duration * position;

        _this.setState({
          progress
        });

        if (position === duration) {
          _this.pauseAudio();
        }
      }
    });
    _this.state = {
      open: false,
      playing: false,
      progress: 0,
      updateProgress: null
    };
    _this.audioRef = React.createRef();
    return _this;
  }

  (0, _createClass2.default)(Audio, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      window.clearInterval(this.state.updateProgress);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var og = this.props.og;
      var audio = og;
      var url = og.asset_url;
      var image = og.image_url;
      return React.createElement("div", {
        className: "str-chat__audio"
      }, React.createElement("div", {
        className: "str-chat__audio__wrapper"
      }, React.createElement("audio", {
        ref: this.audioRef
      }, React.createElement("source", {
        src: url,
        type: "audio/mp3"
      })), React.createElement("div", {
        className: "str-chat__audio__image"
      }, React.createElement("div", {
        className: "str-chat__audio__image--overlay"
      }, !this.state.playing ? React.createElement("div", {
        onClick: function onClick() {
          return _this2.playAudio();
        },
        className: "str-chat__audio__image--button"
      }, React.createElement("svg", {
        width: "40",
        height: "40",
        viewBox: "0 0 64 64",
        xmlns: "http://www.w3.org/2000/svg"
      }, React.createElement("path", {
        d: "M32 58c14.36 0 26-11.64 26-26S46.36 6 32 6 6 17.64 6 32s11.64 26 26 26zm0 6C14.327 64 0 49.673 0 32 0 14.327 14.327 0 32 0c17.673 0 32 14.327 32 32 0 17.673-14.327 32-32 32zm13.237-28.412L26.135 45.625a3.27 3.27 0 0 1-4.426-1.4 3.319 3.319 0 0 1-.372-1.47L21 23.36c-.032-1.823 1.41-3.327 3.222-3.358a3.263 3.263 0 0 1 1.473.322l19.438 9.36a3.311 3.311 0 0 1 .103 5.905z",
        fillRule: "nonzero"
      }))) : React.createElement("div", {
        onClick: function onClick() {
          return _this2.pauseAudio();
        },
        className: "str-chat__audio__image--button"
      }, React.createElement("svg", {
        width: "40",
        height: "40",
        viewBox: "0 0 64 64",
        xmlns: "http://www.w3.org/2000/svg"
      }, React.createElement("path", {
        d: "M32 58.215c14.478 0 26.215-11.737 26.215-26.215S46.478 5.785 32 5.785 5.785 17.522 5.785 32 17.522 58.215 32 58.215zM32 64C14.327 64 0 49.673 0 32 0 14.327 14.327 0 32 0c17.673 0 32 14.327 32 32 0 17.673-14.327 32-32 32zm-7.412-45.56h2.892a2.17 2.17 0 0 1 2.17 2.17v23.865a2.17 2.17 0 0 1-2.17 2.17h-2.892a2.17 2.17 0 0 1-2.17-2.17V20.61a2.17 2.17 0 0 1 2.17-2.17zm12.293 0h2.893a2.17 2.17 0 0 1 2.17 2.17v23.865a2.17 2.17 0 0 1-2.17 2.17h-2.893a2.17 2.17 0 0 1-2.17-2.17V20.61a2.17 2.17 0 0 1 2.17-2.17z",
        fillRule: "nonzero"
      })))), React.createElement("img", {
        src: image,
        alt: "".concat(og.description)
      })), React.createElement("div", {
        className: "str-chat__audio__content"
      }, React.createElement("span", {
        className: "str-chat__audio__content--title"
      }, React.createElement("strong", null, og.title)), React.createElement("span", {
        className: "str-chat__audio__content--subtitle"
      }, og.text), React.createElement("div", {
        className: "str-chat__audio__content--progress"
      }, React.createElement("div", {
        style: {
          width: "".concat(this.state.progress, "%")
        }
      })))));
    }
  }]);
  return Audio;
}(React.Component);

exports.Audio = Audio;
(0, _defineProperty2.default)(Audio, "propTypes", {
  og: _propTypes.default.object
});
