// stream-chat-react/dist/Avatar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Avatar = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Avatar = function (_React$PureComponent) {
  (0, _inherits2.default)(Avatar, _React$PureComponent);

  var _super = _createSuper(Avatar);

  function Avatar() {
    var _this;

    (0, _classCallCheck2.default)(this, Avatar);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      errored: false,
      loaded: false
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getInitials", function (name) {
      return name ? name.split(' ').slice(0, 1).map(function (name) {
        return name.charAt(0);
      }) : null;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onLoad", function () {
      _this.setState({
        loaded: true
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onError", function () {
      _this.setState({
        errored: true
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Avatar, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.image !== this.props.image) {
        this.setState({
          loaded: false,
          errored: false
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          size = _this$props.size,
          name = _this$props.name,
          shape = _this$props.shape,
          image = _this$props.image;
      var initials = this.getInitials(name);
      return _react.default.createElement("div", {
        className: "str-chat__avatar str-chat__avatar--".concat(shape),
        title: name,
        style: {
          width: size,
          height: size,
          flexBasis: size,
          lineHeight: size + 'px',
          fontSize: size / 2
        }
      }, image && !this.state.errored ? _react.default.createElement("img", {
        src: image,
        alt: initials,
        className: 'str-chat__avatar-image' + (this.state.loaded ? ' str-chat__avatar-image--loaded' : ''),
        style: {
          width: size,
          height: size,
          flexBasis: size,
          objectFit: 'cover'
        },
        onLoad: this.onLoad,
        onError: this.onError
      }) : _react.default.createElement("div", {
        className: "str-chat__avatar-fallback"
      }, initials));
    }
  }]);
  return Avatar;
}(_react.default.PureComponent);

exports.Avatar = Avatar;
(0, _defineProperty2.default)(Avatar, "propTypes", {
  image: _propTypes.default.string,
  name: _propTypes.default.string,
  shape: _propTypes.default.oneOf(['circle', 'rounded', 'square']),
  size: _propTypes.default.number
});
(0, _defineProperty2.default)(Avatar, "defaultProps", {
  size: 32,
  shape: 'circle'
});
