// stream-chat-react/dist/Card.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _SafeAnchor = require("./SafeAnchor");

var _Poweredby_100pxWhite_VertText = _interopRequireDefault(require("../assets/Poweredby_100px-White_VertText.png"));

var _context = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Card = function (_React$PureComponent) {
  (0, _inherits2.default)(Card, _React$PureComponent);

  var _super = _createSuper(Card);

  function Card() {
    var _this;

    (0, _classCallCheck2.default)(this, Card);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "trimUrl", function (url) {
      var trimmedUrl;

      if (url !== undefined && url !== null) {
        trimmedUrl = url.replace(/^(?:https?:\/\/)?(?:www\.)?/i, '').split('/')[0];
      }

      return trimmedUrl;
    });
    return _this;
  }

  (0, _createClass2.default)(Card, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          title_link = _this$props.title_link,
          text = _this$props.text,
          type = _this$props.type,
          image_url = _this$props.image_url,
          thumb_url = _this$props.thumb_url,
          og_scrape_url = _this$props.og_scrape_url,
          t = _this$props.t;
      var image = thumb_url || image_url;

      if (!title && !title_link && !image) {
        return _react.default.createElement("div", {
          className: "str-chat__message-attachment-card str-chat__message-attachment-card--".concat(type)
        }, _react.default.createElement("div", {
          className: "str-chat__message-attachment-card--content"
        }, _react.default.createElement("div", {
          className: "str-chat__message-attachment-card--text"
        }, t('this content could not be displayed'))));
      }

      if (!title_link && !og_scrape_url) {
        return null;
      }

      return _react.default.createElement("div", {
        className: "str-chat__message-attachment-card str-chat__message-attachment-card--".concat(type)
      }, image && _react.default.createElement("div", {
        className: "str-chat__message-attachment-card--header"
      }, _react.default.createElement("img", {
        src: image,
        alt: image
      })), _react.default.createElement("div", {
        className: "str-chat__message-attachment-card--content"
      }, _react.default.createElement("div", {
        className: "str-chat__message-attachment-card--flex"
      }, title && _react.default.createElement("div", {
        className: "str-chat__message-attachment-card--title"
      }, title), text && _react.default.createElement("div", {
        className: "str-chat__message-attachment-card--text"
      }, text), (title_link || og_scrape_url) && _react.default.createElement(_SafeAnchor.SafeAnchor, {
        href: title_link || og_scrape_url,
        target: "_blank",
        rel: "noopener noreferrer",
        className: "str-chat__message-attachment-card--url"
      }, this.trimUrl(title_link || og_scrape_url))), type === 'giphy' && _react.default.createElement("img", {
        className: "str-chat__message-attachment-card__giphy-logo",
        src: _Poweredby_100pxWhite_VertText.default,
        alt: "giphy logo"
      })));
    }
  }]);
  return Card;
}(_react.default.PureComponent);

(0, _defineProperty2.default)(Card, "propTypes", {
  title: _propTypes.default.string,
  title_link: _propTypes.default.string,
  og_scrape_url: _propTypes.default.string,
  image_url: _propTypes.default.string,
  thumb_url: _propTypes.default.string,
  text: _propTypes.default.string
});

var _default = (0, _context.withTranslationContext)(Card);

exports.default = _default;
