// stream-chat-react/dist/ChannelPreviewMessenger.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChannelPreviewMessenger = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Avatar = require("./Avatar");

var _truncate = _interopRequireDefault(require("lodash/truncate"));

var _context = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ChannelPreviewMessenger = function (_PureComponent) {
  (0, _inherits2.default)(ChannelPreviewMessenger, _PureComponent);

  var _super = _createSuper(ChannelPreviewMessenger);

  function ChannelPreviewMessenger() {
    var _this;

    (0, _classCallCheck2.default)(this, ChannelPreviewMessenger);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "channelPreviewButton", _react.default.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onSelectChannel", function () {
      _this.props.setActiveChannel(_this.props.channel, _this.props.watchers);

      _this.channelPreviewButton.current.blur();
    });
    return _this;
  }

  (0, _createClass2.default)(ChannelPreviewMessenger, [{
    key: "render",
    value: function render() {
      var unreadClass = this.props.unread >= 1 ? 'str-chat__channel-preview-messenger--unread' : '';
      var activeClass = this.props.active ? 'str-chat__channel-preview-messenger--active' : '';
      var _this$props = this.props,
          channel = _this$props.channel,
          displayTitle = _this$props.displayTitle,
          displayImage = _this$props.displayImage,
          t = _this$props.t;
      return _react.default.createElement("button", {
        onClick: this.onSelectChannel,
        ref: this.channelPreviewButton,
        className: "str-chat__channel-preview-messenger ".concat(unreadClass, " ").concat(activeClass)
      }, _react.default.createElement("div", {
        className: "str-chat__channel-preview-messenger--left"
      }, _react.default.createElement(_Avatar.Avatar, {
        image: displayImage,
        size: 40
      })), _react.default.createElement("div", {
        className: "str-chat__channel-preview-messenger--right"
      }, _react.default.createElement("div", {
        className: "str-chat__channel-preview-messenger--name"
      }, _react.default.createElement("span", null, displayTitle)), _react.default.createElement("div", {
        className: "str-chat__channel-preview-messenger--last-message"
      }, !channel.state.messages[0] ? t('Nothing yet...') : (0, _truncate.default)(this.props.latestMessage, {
        length: this.props.latestMessageLength
      }))));
    }
  }]);
  return ChannelPreviewMessenger;
}(_react.PureComponent);

exports.ChannelPreviewMessenger = ChannelPreviewMessenger;
(0, _defineProperty2.default)(ChannelPreviewMessenger, "propTypes", {
  channel: _propTypes.default.object.isRequired,
  activeChannel: _propTypes.default.object,
  setActiveChannel: _propTypes.default.func.isRequired,
  watchers: _propTypes.default.object,
  unread: _propTypes.default.number,
  active: _propTypes.default.bool,
  latestMessage: _propTypes.default.string,
  latestMessageLength: _propTypes.default.number,
  displayTitle: _propTypes.default.string,
  displayImage: _propTypes.default.string
});
(0, _defineProperty2.default)(ChannelPreviewMessenger, "defaultProps", {
  latestMessageLength: 14
});
exports.ChannelPreviewMessenger = ChannelPreviewMessenger = (0, _context.withTranslationContext)(ChannelPreviewMessenger);
