// stream-chat-react/dist/ChannelListMessenger.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChannelListMessenger = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _LoadingChannels = require("./LoadingChannels");

var _ChatDown = require("./ChatDown");

var _context = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ChannelListMessenger = function (_PureComponent) {
  (0, _inherits2.default)(ChannelListMessenger, _PureComponent);

  var _super = _createSuper(ChannelListMessenger);

  function ChannelListMessenger() {
    (0, _classCallCheck2.default)(this, ChannelListMessenger);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ChannelListMessenger, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          error = _this$props.error,
          loading = _this$props.loading,
          LoadingErrorIndicator = _this$props.LoadingErrorIndicator,
          LoadingIndicator = _this$props.LoadingIndicator;

      if (error) {
        return _react.default.createElement(LoadingErrorIndicator, {
          type: "Connection Error"
        });
      } else if (loading) {
        return _react.default.createElement(LoadingIndicator, null);
      } else {
        return _react.default.createElement("div", {
          className: "str-chat__channel-list-messenger"
        }, _react.default.createElement("div", {
          className: "str-chat__channel-list-messenger__main"
        }, this.props.children));
      }
    }
  }]);
  return ChannelListMessenger;
}(_react.PureComponent);

exports.ChannelListMessenger = ChannelListMessenger;
(0, _defineProperty2.default)(ChannelListMessenger, "propTypes", {
  loading: _propTypes.default.bool,
  error: _propTypes.default.bool,
  LoadingIndicator: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  LoadingErrorIndicator: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func])
});
(0, _defineProperty2.default)(ChannelListMessenger, "defaultProps", {
  error: false,
  LoadingIndicator: _LoadingChannels.LoadingChannels,
  LoadingErrorIndicator: _ChatDown.ChatDown
});
exports.ChannelListMessenger = ChannelListMessenger = (0, _context.withChatContext)(ChannelListMessenger);
