// stream-chat-react/dist/ChannelPreviewLastMessage.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChannelPreviewLastMessage = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _Avatar = require("./Avatar");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _truncate = _interopRequireDefault(require("lodash/truncate"));

var _context = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ChannelPreviewLastMessage = function (_PureComponent) {
  (0, _inherits2.default)(ChannelPreviewLastMessage, _PureComponent);

  var _super = _createSuper(ChannelPreviewLastMessage);

  function ChannelPreviewLastMessage() {
    var _this;

    (0, _classCallCheck2.default)(this, ChannelPreviewLastMessage);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "channelPreviewButton", _react.default.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onSelectChannel", function () {
      _this.props.setActiveChannel(_this.props.channel, _this.props.watchers);

      _this.channelPreviewButton.current.blur();
    });
    return _this;
  }

  (0, _createClass2.default)(ChannelPreviewLastMessage, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          t = _this$props.t,
          displayTitle = _this$props.displayTitle,
          displayImage = _this$props.displayImage;
      var unreadClass = this.props.unread >= 1 ? 'str-chat__channel-preview--unread' : '';
      var activeClass = this.props.active ? 'str-chat__channel-preview--active' : '';
      return _react.default.createElement("div", {
        className: "str-chat__channel-preview ".concat(unreadClass, " ").concat(activeClass)
      }, _react.default.createElement("button", {
        onClick: this.onSelectChannel,
        ref: this.channelPreviewButton
      }, this.props.unread >= 1 && _react.default.createElement("div", {
        className: "str-chat__channel-preview--dot"
      }), _react.default.createElement(_Avatar.Avatar, {
        image: displayImage
      }), _react.default.createElement("div", {
        className: "str-chat__channel-preview-info"
      }, _react.default.createElement("span", {
        className: "str-chat__channel-preview-title"
      }, displayTitle), _react.default.createElement("span", {
        className: "str-chat__channel-preview-last-message"
      }, !this.props.channel.state.messages[0] ? t('Nothing yet...') : (0, _truncate.default)(this.props.latestMessage, {
        length: this.props.latestMessageLength
      })), this.props.unread >= 1 && _react.default.createElement("span", {
        className: "str-chat__channel-preview-unread-count"
      }, this.props.unread))));
    }
  }]);
  return ChannelPreviewLastMessage;
}(_react.PureComponent);

exports.ChannelPreviewLastMessage = ChannelPreviewLastMessage;
(0, _defineProperty2.default)(ChannelPreviewLastMessage, "propTypes", {
  channel: _propTypes.default.object.isRequired,
  activeChannel: _propTypes.default.object,
  setActiveChannel: _propTypes.default.func.isRequired,
  watchers: _propTypes.default.object,
  unread: _propTypes.default.number,
  active: _propTypes.default.bool,
  latestMessage: _propTypes.default.string,
  latestMessageLength: _propTypes.default.number,
  displayTitle: _propTypes.default.string,
  displayImage: _propTypes.default.string
});
(0, _defineProperty2.default)(ChannelPreviewLastMessage, "defaultProps", {
  latestMessageLength: 20
});
exports.ChannelPreviewLastMessage = ChannelPreviewLastMessage = (0, _context.withTranslationContext)(ChannelPreviewLastMessage);
