// stream-chat-react/dist/ChannelPreviewCountOnly.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChannelPreviewCountOnly = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ChannelPreviewCountOnly = function (_PureComponent) {
  (0, _inherits2.default)(ChannelPreviewCountOnly, _PureComponent);

  var _super = _createSuper(ChannelPreviewCountOnly);

  function ChannelPreviewCountOnly() {
    (0, _classCallCheck2.default)(this, ChannelPreviewCountOnly);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ChannelPreviewCountOnly, [{
    key: "render",
    value: function render() {
      var displayTitle = this.props.displayTitle;
      var unreadClass = this.props.unread >= 1 ? 'unread' : '';
      return _react.default.createElement("div", {
        className: unreadClass
      }, _react.default.createElement("button", {
        onClick: this.props.setActiveChannel.bind(this, this.props.channel)
      }, ' ', displayTitle, " ", _react.default.createElement("span", null, this.props.unread)));
    }
  }]);
  return ChannelPreviewCountOnly;
}(_react.PureComponent);

exports.ChannelPreviewCountOnly = ChannelPreviewCountOnly;
(0, _defineProperty2.default)(ChannelPreviewCountOnly, "propTypes", {
  channel: _propTypes.default.object.isRequired,
  activeChannel: _propTypes.default.object,
  setActiveChannel: _propTypes.default.func.isRequired,
  watchers: _propTypes.default.object,
  unread: _propTypes.default.number,
  displayTitle: _propTypes.default.string,
  displayImage: _propTypes.default.string
});
