// stream-chat-react/dist/ChatDown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChatDown = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _LoadingChannels = require("./LoadingChannels");

var _strChat__connectionError = _interopRequireDefault(require("../assets/str-chat__connection-error.svg"));

var _context = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ChatDown = function (_React$PureComponent) {
  (0, _inherits2.default)(ChatDown, _React$PureComponent);

  var _super = _createSuper(ChatDown);

  function ChatDown() {
    (0, _classCallCheck2.default)(this, ChatDown);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ChatDown, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          image = _this$props.image,
          type = _this$props.type,
          text = _this$props.text,
          t = _this$props.t;
      return _react.default.createElement("div", {
        className: "str-chat__down"
      }, _react.default.createElement(_LoadingChannels.LoadingChannels, null), _react.default.createElement("div", {
        className: "str-chat__down-main"
      }, _react.default.createElement("img", {
        src: image
      }), _react.default.createElement("h1", null, type), _react.default.createElement("h3", null, text || t('Error connecting to chat, refresh the page to try again.'))));
    }
  }]);
  return ChatDown;
}(_react.default.PureComponent);

exports.ChatDown = ChatDown;
(0, _defineProperty2.default)(ChatDown, "propTypes", {
  image: _propTypes.default.string,
  type: _propTypes.default.string,
  text: _propTypes.default.string
});
(0, _defineProperty2.default)(ChatDown, "defaultProps", {
  image: _strChat__connectionError.default,
  type: 'Error'
});
exports.ChatDown = ChatDown = (0, _context.withTranslationContext)(ChatDown);
