// stream-chat-react/dist/context.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withChatContext = withChatContext;
exports.withChannelContext = withChannelContext;
exports.withTranslationContext = withTranslationContext;
exports.TranslationContext = exports.ChannelContext = exports.ChatContext = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _dayjs = _interopRequireDefault(require("dayjs"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var ChatContext = _react.default.createContext({
  client: null
});

exports.ChatContext = ChatContext;

function withChatContext(OriginalComponent) {
  var ContextAwareComponent = function ContextComponent(props) {
    return _react.default.createElement(ChatContext.Consumer, null, function (context) {
      var mergedProps = _objectSpread({}, context, {}, props);

      return _react.default.createElement(OriginalComponent, mergedProps);
    });
  };

  ContextAwareComponent.displayName = OriginalComponent.displayName || OriginalComponent.name || 'Component';
  ContextAwareComponent.displayName = ContextAwareComponent.displayName.replace('Base', '');
  return ContextAwareComponent;
}

var ChannelContext = _react.default.createContext({});

exports.ChannelContext = ChannelContext;

function withChannelContext(OriginalComponent) {
  var ContextAwareComponent = function ContextComponent(props) {
    return _react.default.createElement(ChannelContext.Consumer, null, function (channelContext) {
      return _react.default.createElement(OriginalComponent, (0, _extends2.default)({}, channelContext, props));
    });
  };

  ContextAwareComponent.displayName = OriginalComponent.displayName || OriginalComponent.name || 'Component';
  ContextAwareComponent.displayName = ContextAwareComponent.displayName.replace('Base', '');
  return ContextAwareComponent;
}

var TranslationContext = _react.default.createContext({
  t: function t(msg) {
    return msg;
  },
  tDateTimeParser: function tDateTimeParser(input) {
    return (0, _dayjs.default)(input);
  }
});

exports.TranslationContext = TranslationContext;

function withTranslationContext(OriginalComponent) {
  var ContextAwareComponent = function ContextComponent(props) {
    return _react.default.createElement(TranslationContext.Consumer, null, function (translationContext) {
      return _react.default.createElement(OriginalComponent, (0, _extends2.default)({}, translationContext, props));
    });
  };

  ContextAwareComponent.displayName = OriginalComponent.displayName || OriginalComponent.name || 'Component';
  ContextAwareComponent.displayName = ContextAwareComponent.displayName.replace('Base', '');
  return ContextAwareComponent;
}
