// stream-chat-react/dist/DateSeparator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DateSeparator = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _context = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var DateSeparator = function (_React$PureComponent) {
  (0, _inherits2.default)(DateSeparator, _React$PureComponent);

  var _super = _createSuper(DateSeparator);

  function DateSeparator() {
    (0, _classCallCheck2.default)(this, DateSeparator);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(DateSeparator, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          position = _this$props.position,
          tDateTimeParser = _this$props.tDateTimeParser;

      if (!Date.parse(this.props.date)) {
        return null;
      }

      return _react.default.createElement("div", {
        className: "str-chat__date-separator"
      }, (position === 'right' || position === 'center') && _react.default.createElement("hr", {
        className: "str-chat__date-separator-line"
      }), _react.default.createElement("div", {
        className: "str-chat__date-separator-date"
      }, this.props.formatDate ? this.props.formatDate(this.props.date) : tDateTimeParser(this.props.date.toISOString()).calendar()), (position === 'left' || position === 'center') && _react.default.createElement("hr", {
        className: "str-chat__date-separator-line"
      }));
    }
  }]);
  return DateSeparator;
}(_react.default.PureComponent);

exports.DateSeparator = DateSeparator;
(0, _defineProperty2.default)(DateSeparator, "propTypes", {
  date: _propTypes.default.instanceOf(Date),
  position: _propTypes.default.oneOf(['left', 'center', 'right']),
  formatDate: _propTypes.default.func
});
(0, _defineProperty2.default)(DateSeparator, "defaultProps", {
  position: 'right'
});
exports.DateSeparator = DateSeparator = (0, _context.withTranslationContext)(DateSeparator);
