// stream-chat-react/dist/Gallery.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Gallery = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactImages = _interopRequireWildcard(require("react-images"));

var _context = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Gallery = function (_React$PureComponent) {
  (0, _inherits2.default)(Gallery, _React$PureComponent);

  var _super = _createSuper(Gallery);

  function Gallery() {
    var _this;

    (0, _classCallCheck2.default)(this, Gallery);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      modalIsOpen: false,
      currentIndex: 0
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "toggleModal", function (index) {
      _this.setState(function (state) {
        return {
          modalIsOpen: !state.modalIsOpen,
          currentIndex: index
        };
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Gallery, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          images = _this$props.images,
          t = _this$props.t;
      var formattedArray = images.map(function (image) {
        return {
          src: image.image_url || image.thumb_url
        };
      });
      return _react.default.createElement("div", {
        className: "str-chat__gallery"
      }, images.slice(0, 3).map(function (image, i) {
        return _react.default.createElement("div", {
          className: "str-chat__gallery-image",
          key: "gallery-image-".concat(i),
          onClick: function onClick() {
            return _this2.toggleModal(i);
          }
        }, _react.default.createElement("img", {
          src: image.image_url || image.thumb_url
        }));
      }), images.length > 3 && _react.default.createElement("div", {
        className: "str-chat__gallery-placeholder",
        style: {
          backgroundImage: "url(".concat(images[3].image_url, ")")
        },
        onClick: function onClick() {
          return _this2.toggleModal(3);
        }
      }, _react.default.createElement("p", null, t('{{ imageCount }} more', {
        imageCount: images.length - 3
      }))), _react.default.createElement(_reactImages.ModalGateway, null, this.state.modalIsOpen ? _react.default.createElement(_reactImages.Modal, {
        onClose: this.toggleModal,
        closeOnBackdropClick: true
      }, _react.default.createElement(_reactImages.default, {
        views: formattedArray,
        currentIndex: this.state.currentIndex
      })) : null));
    }
  }]);
  return Gallery;
}(_react.default.PureComponent);

exports.Gallery = Gallery;
(0, _defineProperty2.default)(Gallery, "propTypes", {
  images: _propTypes.default.arrayOf(_propTypes.default.shape({
    image_url: _propTypes.default.string,
    thumb_url: _propTypes.default.string
  }))
});
exports.Gallery = Gallery = (0, _context.withTranslationContext)(Gallery);
