// stream-chat-react/dist/Image.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Image = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactImages = _interopRequireWildcard(require("react-images"));

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Image = function (_React$PureComponent) {
  (0, _inherits2.default)(Image, _React$PureComponent);

  var _super = _createSuper(Image);

  function Image() {
    var _this;

    (0, _classCallCheck2.default)(this, Image);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      modalIsOpen: false,
      currentIndex: 0
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "toggleModal", function () {
      _this.setState(function (state) {
        return {
          modalIsOpen: !state.modalIsOpen
        };
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Image, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          image_url = _this$props.image_url,
          thumb_url = _this$props.thumb_url,
          fallback = _this$props.fallback;
      var formattedArray = [{
        src: image_url || thumb_url
      }];
      return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("img", {
        className: "str-chat__message-attachment--img",
        onClick: this.toggleModal,
        src: thumb_url || image_url,
        alt: fallback
      }), _react.default.createElement(_reactImages.ModalGateway, null, this.state.modalIsOpen ? _react.default.createElement(_reactImages.Modal, {
        onClose: this.toggleModal
      }, _react.default.createElement(_reactImages.default, {
        views: formattedArray
      })) : null));
    }
  }]);
  return Image;
}(_react.default.PureComponent);

exports.Image = Image;
(0, _defineProperty2.default)(Image, "propTypes", {
  image_url: _propTypes.default.string,
  thumb_url: _propTypes.default.string,
  fallback: _propTypes.default.string
});
