// stream-chat-react/dist/InfiniteScrollPaginator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InfiniteScrollPaginator = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactFileUtils = require("react-file-utils");

var _InfiniteScroll = _interopRequireDefault(require("./InfiniteScroll"));

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var InfiniteScrollPaginator = function (_React$Component) {
  (0, _inherits2.default)(InfiniteScrollPaginator, _React$Component);

  var _super = _createSuper(InfiniteScrollPaginator);

  function InfiniteScrollPaginator() {
    (0, _classCallCheck2.default)(this, InfiniteScrollPaginator);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(InfiniteScrollPaginator, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          LoadingIndicator = _this$props.LoadingIndicator,
          loadNextPage = _this$props.loadNextPage,
          hasNextPage = _this$props.hasNextPage,
          refreshing = _this$props.refreshing,
          reverse = _this$props.reverse,
          threshold = _this$props.threshold;
      return _react.default.createElement(_InfiniteScroll.default, {
        loadMore: loadNextPage,
        hasMore: hasNextPage,
        isLoading: refreshing,
        isReverse: reverse,
        threshold: threshold,
        useWindow: false,
        loader: _react.default.createElement("div", {
          className: "str-chat__infinite-scroll-paginator",
          key: "loadingindicator"
        }, _react.default.createElement(LoadingIndicator, null))
      }, this.props.children);
    }
  }]);
  return InfiniteScrollPaginator;
}(_react.default.Component);

exports.InfiniteScrollPaginator = InfiniteScrollPaginator;
(0, _defineProperty2.default)(InfiniteScrollPaginator, "propTypes", {
  loadNextPage: _propTypes.default.func,
  hasNextPage: _propTypes.default.bool,
  refreshing: _propTypes.default.bool,
  reverse: _propTypes.default.bool,
  threshold: _propTypes.default.number,
  LoadingIndicator: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func])
});
(0, _defineProperty2.default)(InfiniteScrollPaginator, "defaultProps", {
  LoadingIndicator: _reactFileUtils.LoadingIndicator
});
