// stream-chat-react/dist/LoadingIndicator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoadingIndicator = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _strChat__loadingIndicator = _interopRequireDefault(require("../assets/str-chat__loading-indicator.svg"));

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var LoadingIndicator = function (_React$PureComponent) {
  (0, _inherits2.default)(LoadingIndicator, _React$PureComponent);

  var _super = _createSuper(LoadingIndicator);

  function LoadingIndicator() {
    var _this;

    (0, _classCallCheck2.default)(this, LoadingIndicator);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "stopRef", _react.default.createRef());
    return _this;
  }

  (0, _createClass2.default)(LoadingIndicator, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          size = _this$props.size,
          color = _this$props.color;
      return _react.default.createElement("div", {
        className: 'str-chat__loading-indicator ' + color,
        style: {
          width: size,
          height: size
        }
      }, _react.default.createElement("svg", {
        width: size,
        height: size,
        viewBox: "0 0 30 30",
        xmlns: "http://www.w3.org/2000/svg"
      }, _react.default.createElement("defs", null, _react.default.createElement("linearGradient", {
        x1: "50%",
        y1: "0%",
        x2: "50%",
        y2: "100%",
        id: "a"
      }, _react.default.createElement("stop", {
        stopColor: "#FFF",
        stopOpacity: "0",
        offset: "0%"
      }), _react.default.createElement("stop", {
        ref: this.stopRef,
        offset: "100%",
        stopColor: color,
        stopOpacity: "1",
        style: {
          stopColor: color
        }
      }))), _react.default.createElement("path", {
        d: "M2.518 23.321l1.664-1.11A12.988 12.988 0 0 0 15 28c7.18 0 13-5.82 13-13S22.18 2 15 2V0c8.284 0 15 6.716 15 15 0 8.284-6.716 15-15 15-5.206 0-9.792-2.652-12.482-6.679z",
        fill: "url(#a)",
        fillRule: "evenodd"
      })));
    }
  }]);
  return LoadingIndicator;
}(_react.default.PureComponent);

exports.LoadingIndicator = LoadingIndicator;
(0, _defineProperty2.default)(LoadingIndicator, "propTypes", {
  size: _propTypes.default.number,
  color: _propTypes.default.string
});
(0, _defineProperty2.default)(LoadingIndicator, "defaultProps", {
  size: 15,
  color: '#006CFF'
});
