// stream-chat-react/dist/MessageCommerce.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageCommerce = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _Attachment = require("./Attachment");

var _ReactionsList = require("./ReactionsList");

var _Avatar = require("./Avatar");

var _Gallery = require("./Gallery");

var _ReactionSelector = require("./ReactionSelector");

var _MessageRepliesCountButton = require("./MessageRepliesCountButton");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

var _context = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var MessageCommerce = function (_PureComponent) {
  (0, _inherits2.default)(MessageCommerce, _PureComponent);

  var _super = _createSuper(MessageCommerce);

  function MessageCommerce() {
    var _this;

    (0, _classCallCheck2.default)(this, MessageCommerce);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      isFocused: false,
      actionsBoxOpen: false,
      showDetailedReactions: false
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "messageActionsRef", _react.default.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "reactionSelectorRef", _react.default.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_onClickOptionsAction", function () {
      _this.setState({
        actionsBoxOpen: true
      }, function () {
        return document.addEventListener('click', _this.hideOptions, false);
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_hideOptions", function () {
      _this.setState({
        actionsBoxOpen: false
      });

      document.removeEventListener('click', _this.hideOptions, false);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_clickReactionList", function () {
      _this.setState(function () {
        return {
          showDetailedReactions: true
        };
      }, function () {
        document.addEventListener('click', _this._closeDetailedReactions);
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_closeDetailedReactions", function (e) {
      if (!_this.reactionSelectorRef.current.reactionSelector.current.contains(e.target)) {
        _this.setState(function () {
          return {
            showDetailedReactions: false
          };
        }, function () {
          document.removeEventListener('click', _this._closeDetailedReactions);
        });
      } else {
        return {};
      }
    });
    return _this;
  }

  (0, _createClass2.default)(MessageCommerce, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (!this.props.message.deleted_at) {
        document.removeEventListener('click', this._closeDetailedReactions);
      }
    }
  }, {
    key: "isMine",
    value: function isMine() {
      return !this.props.isMyMessage(this.props.message);
    }
  }, {
    key: "renderOptions",
    value: function renderOptions() {
      if (this.props.message.type === 'error' || this.props.message.type === 'system' || this.props.message.type === 'ephemeral' || this.props.message.status === 'sending' || this.props.message.status === 'failed' || !this.props.channelConfig.reactions || this.props.initialMessage) {
        return;
      }

      return _react.default.createElement("div", {
        className: "str-chat__message-commerce__actions"
      }, _react.default.createElement("div", {
        className: "str-chat__message-commerce__actions__action str-chat__message-commerce__actions__action--reactions",
        onClick: this._clickReactionList
      }, _react.default.createElement("svg", {
        width: "14",
        height: "14",
        xmlns: "http://www.w3.org/2000/svg"
      }, _react.default.createElement("path", {
        d: "M11.108 8.05a.496.496 0 0 1 .212.667C10.581 10.147 8.886 11 7 11c-1.933 0-3.673-.882-4.33-2.302a.497.497 0 0 1 .9-.417C4.068 9.357 5.446 10 7 10c1.519 0 2.869-.633 3.44-1.738a.495.495 0 0 1 .668-.212zm.792-1.826a.477.477 0 0 1-.119.692.541.541 0 0 1-.31.084.534.534 0 0 1-.428-.194c-.106-.138-.238-.306-.539-.306-.298 0-.431.168-.54.307A.534.534 0 0 1 9.538 7a.544.544 0 0 1-.31-.084.463.463 0 0 1-.117-.694c.33-.423.742-.722 1.394-.722.653 0 1.068.3 1.396.724zm-7 0a.477.477 0 0 1-.119.692.541.541 0 0 1-.31.084.534.534 0 0 1-.428-.194c-.106-.138-.238-.306-.539-.306-.299 0-.432.168-.54.307A.533.533 0 0 1 2.538 7a.544.544 0 0 1-.31-.084.463.463 0 0 1-.117-.694c.33-.423.742-.722 1.394-.722.653 0 1.068.3 1.396.724zM7 0a7 7 0 1 1 0 14A7 7 0 0 1 7 0zm4.243 11.243A5.96 5.96 0 0 0 13 7a5.96 5.96 0 0 0-1.757-4.243A5.96 5.96 0 0 0 7 1a5.96 5.96 0 0 0-4.243 1.757A5.96 5.96 0 0 0 1 7a5.96 5.96 0 0 0 1.757 4.243A5.96 5.96 0 0 0 7 13a5.96 5.96 0 0 0 4.243-1.757z",
        fillRule: "evenodd"
      }))));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          message = _this$props.message,
          groupStyles = _this$props.groupStyles,
          Attachment = _this$props.Attachment,
          handleReaction = _this$props.handleReaction,
          handleAction = _this$props.handleAction,
          actionsEnabled = _this$props.actionsEnabled,
          onMentionsHoverMessage = _this$props.onMentionsHoverMessage,
          onMentionsClickMessage = _this$props.onMentionsClickMessage,
          unsafeHTML = _this$props.unsafeHTML,
          threadList = _this$props.threadList,
          handleOpenThread = _this$props.handleOpenThread,
          t = _this$props.t,
          tDateTimeParser = _this$props.tDateTimeParser;
      var when = tDateTimeParser(message.created_at).format('LT');
      var messageClasses = this.isMine() ? 'str-chat__message-commerce str-chat__message-commerce--left' : 'str-chat__message-commerce str-chat__message-commerce--right';
      var hasAttachment = Boolean(message.attachments && message.attachments.length);
      var images = hasAttachment && message.attachments.filter(function (item) {
        return item.type === 'image';
      });
      var hasReactions = Boolean(message.latest_reactions && message.latest_reactions.length);

      if (message.type === 'message.read' || message.deleted_at || message.type === 'message.date') {
        return null;
      }

      if (message.deleted_at) {
        return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("span", {
          key: message.id,
          className: "".concat(messageClasses, " str-chat__message--deleted")
        }, t('This message was deleted...')), _react.default.createElement("div", {
          className: "clearfix"
        }));
      }

      return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
        key: message.id,
        className: "\n\t\t\t\t\t\t".concat(messageClasses, "\n\t\t\t\t\t\tstr-chat__message-commerce--").concat(message.type, "\n\t\t\t\t\t\t").concat(message.text ? 'str-chat__message-commerce--has-text' : 'str-chat__message-commerce--has-no-text', "\n\t\t\t\t\t\t").concat(hasAttachment ? 'str-chat__message-commerce--has-attachment' : '', "\n\t\t\t\t\t\t").concat(hasReactions ? 'str-chat__message-commerce--with-reactions' : '', "\n\t\t\t\t\t\t", "str-chat__message-commerce--".concat(groupStyles[0]), "\n\t\t\t\t\t").trim(),
        onMouseLeave: this._hideOptions,
        ref: this.messageRef
      }, (groupStyles[0] === 'bottom' || groupStyles[0] === 'single') && _react.default.createElement(_Avatar.Avatar, {
        image: message.user.image,
        size: 32,
        name: message.user.name || message.user.id
      }), _react.default.createElement("div", {
        className: "str-chat__message-commerce-inner"
      }, !message.text && _react.default.createElement(_react.default.Fragment, null, this.renderOptions(), hasReactions && !this.state.showDetailedReactions && _react.default.createElement(_ReactionsList.ReactionsList, {
        reactions: message.latest_reactions,
        reaction_counts: message.reaction_counts,
        onClick: this._clickReactionList
      }), this.state.showDetailedReactions && _react.default.createElement(_ReactionSelector.ReactionSelector, {
        reverse: false,
        handleReaction: handleReaction,
        actionsEnabled: actionsEnabled,
        detailedView: true,
        reaction_counts: message.reaction_counts,
        latest_reactions: message.latest_reactions,
        ref: this.reactionSelectorRef
      })), hasAttachment && images.length <= 1 && message.attachments.map(function (attachment, index) {
        return _react.default.createElement(Attachment, {
          key: "".concat(message.id, "-").concat(index),
          attachment: attachment,
          actionHandler: handleAction
        });
      }), images.length > 1 && _react.default.createElement(_Gallery.Gallery, {
        images: images
      }), message.text && _react.default.createElement("div", {
        className: "str-chat__message-commerce-text"
      }, _react.default.createElement("div", {
        className: "str-chat__message-commerce-text-inner\n\t\t\t\t\t\t\t\t\t".concat(hasAttachment ? 'str-chat__message-commerce-text-inner--has-attachment' : '', "\n\t\t\t\t\t\t\t\t\t").concat((0, _utils.isOnlyEmojis)(message.text) ? 'str-chat__message-commerce-text-inner--is-emoji' : '', "\n                ").trim(),
        onMouseOver: onMentionsHoverMessage,
        onClick: onMentionsClickMessage
      }, message.type === 'error' && _react.default.createElement("div", {
        className: "str-chat__commerce-message--error-message"
      }, t('Error · Unsent')), unsafeHTML ? _react.default.createElement("div", {
        dangerouslySetInnerHTML: {
          __html: message.html
        }
      }) : (0, _utils.renderText)(message), hasReactions && !this.state.showDetailedReactions && _react.default.createElement(_ReactionsList.ReactionsList, {
        reverse: true,
        reactions: message.latest_reactions,
        reaction_counts: message.reaction_counts,
        onClick: this._clickReactionList
      }), this.state.showDetailedReactions && _react.default.createElement(_ReactionSelector.ReactionSelector, {
        reverse: false,
        handleReaction: handleReaction,
        actionsEnabled: actionsEnabled,
        detailedView: true,
        reaction_counts: message.reaction_counts,
        latest_reactions: message.latest_reactions,
        ref: this.reactionSelectorRef
      })), message.text && this.renderOptions()), !threadList && _react.default.createElement("div", {
        className: "str-chat__message-commerce-reply-button"
      }, _react.default.createElement(_MessageRepliesCountButton.MessageRepliesCountButton, {
        onClick: handleOpenThread,
        reply_count: message.reply_count
      })), _react.default.createElement("div", {
        className: "str-chat__message-commerce-data"
      }, this.isMine() ? _react.default.createElement("span", {
        className: "str-chat__message-commerce-name"
      }, message.user.name || message.user.id) : null, _react.default.createElement("span", {
        className: "str-chat__message-commerce-timestamp"
      }, when)))));
    }
  }]);
  return MessageCommerce;
}(_react.PureComponent);

exports.MessageCommerce = MessageCommerce;
(0, _defineProperty2.default)(MessageCommerce, "propTypes", {
  message: _propTypes.default.object,
  Attachment: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  Message: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func, _propTypes.default.object]).isRequired,
  unsafeHTML: _propTypes.default.bool,
  initialMessage: _propTypes.default.bool,
  channelConfig: _propTypes.default.object,
  threadList: _propTypes.default.bool,
  handleOpenThread: _propTypes.default.func,
  isMyMessage: _propTypes.default.func,
  getMessageActions: _propTypes.default.func,
  handleReaction: _propTypes.default.func,
  actionsEnabled: _propTypes.default.bool,
  handleAction: _propTypes.default.func,
  onMentionsHoverMessage: _propTypes.default.func,
  onMentionsClickMessage: _propTypes.default.func,
  groupStyles: _propTypes.default.array
});
(0, _defineProperty2.default)(MessageCommerce, "defaultProps", {
  Attachment: _Attachment.Attachment
});
exports.MessageCommerce = MessageCommerce = (0, _context.withTranslationContext)(MessageCommerce);
