// stream-chat-react/dist/MessageActions.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageActions = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _MessageActionsBox = require("./MessageActionsBox");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var MessageActions = function (_PureComponent) {
  (0, _inherits2.default)(MessageActions, _PureComponent);

  var _super = _createSuper(MessageActions);

  function MessageActions(props) {
    var _this;

    (0, _classCallCheck2.default)(this, MessageActions);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_openActionBox", function () {
      _this.setState({
        actionBox: true
      }, function () {
        document.addEventListener('click', _this._closeActionBox);
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_closeActionBox", function () {
      _this.setState({
        actionBox: false
      }, function () {
        document.removeEventListener('click', _this._closeActionBox);
      });
    });
    _this.state = {
      actionBox: false,
      reactionBox: false
    };
    _this.reactionsBox = _react.default.createRef();
    _this.actionsRef = _react.default.createRef();
    return _this;
  }

  (0, _createClass2.default)(MessageActions, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener('click', this._closeActionBox);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return _react.default.createElement("div", {
        ref: this.actionsRef,
        className: "str-chat__message-actions"
      }, _react.default.createElement("div", {
        className: "str-chat__message-actions-reactions",
        onClick: this.props.onClickReact
      }, _react.default.createElement("svg", {
        width: "20",
        height: "18",
        viewBox: "0 0 20 18",
        xmlns: "http://www.w3.org/2000/svg"
      }, _react.default.createElement("path", {
        d: "M16.5 4.5H15a.5.5 0 1 1 0-1h1.5V2a.5.5 0 1 1 1 0v1.5H19a.5.5 0 1 1 0 1h-1.5V6a.5.5 0 1 1-1 0V4.5zM9 13c-1.773 0-3.297-.82-4-2h8c-.703 1.18-2.227 2-4 2zm4.057-11.468a.5.5 0 1 1-.479.878A7.45 7.45 0 0 0 9 1.5C4.865 1.5 1.5 4.865 1.5 9s3.365 7.5 7.5 7.5 7.5-3.365 7.5-7.5c0-.315-.02-.628-.058-.937a.5.5 0 1 1 .992-.124c.044.35.066.704.066 1.06 0 4.688-3.813 8.501-8.5 8.501C4.313 17.5.5 13.687.5 9 .5 4.312 4.313.5 9 .5a8.45 8.45 0 0 1 4.057 1.032zM7.561 5.44a1.5 1.5 0 1 1-2.123 2.122 1.5 1.5 0 0 1 2.123-2.122zm5 0a1.5 1.5 0 1 1-2.122 2.122 1.5 1.5 0 0 1 2.122-2.122z",
        fillRule: "evenodd"
      }))), _react.default.createElement("div", {
        className: "str-chat__message-actions-options",
        onClick: function onClick() {
          return _this2._openActionBox();
        }
      }, _react.default.createElement("svg", {
        width: "11",
        height: "3",
        viewBox: "0 0 11 3",
        xmlns: "http://www.w3.org/2000/svg"
      }, _react.default.createElement("path", {
        d: "M1.5 3a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm4 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z",
        fillRule: "nonzero"
      }))), _react.default.createElement(_MessageActionsBox.MessageActionsBox, (0, _extends2.default)({}, this.props, {
        open: this.state.actionBox
      })));
    }
  }]);
  return MessageActions;
}(_react.PureComponent);

exports.MessageActions = MessageActions;
(0, _defineProperty2.default)(MessageActions, "propTypes", {
  onClickReact: _propTypes.default.func,
  open: _propTypes.default.bool.isRequired,
  Message: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func, _propTypes.default.object]).isRequired,
  mine: _propTypes.default.bool,
  messageListRect: _propTypes.default.object,
  handleFlag: _propTypes.default.func,
  handleMute: _propTypes.default.func,
  handleEdit: _propTypes.default.func,
  handleDelete: _propTypes.default.func,
  getMessageActions: _propTypes.default.func
});
MessageActions.propTypes = {
  Message: _propTypes.default.object.isRequired
};
