// stream-chat-react/dist/MessageActionsBox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageActionsBox = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

var _context2 = require("../context");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var MessageActionsBox = function (_React$Component) {
  (0, _inherits2.default)(MessageActionsBox, _React$Component);

  var _super = _createSuper(MessageActionsBox);

  function MessageActionsBox() {
    var _this;

    (0, _classCallCheck2.default)(this, MessageActionsBox);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionsBoxRef", _react.default.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      reverse: false,
      rect: null
    });
    return _this;
  }

  (0, _createClass2.default)(MessageActionsBox, [{
    key: "componentDidMount",
    value: function componentDidMount() {}
  }, {
    key: "componentDidUpdate",
    value: function () {
      var _componentDidUpdate = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(prevProps) {
        var ml;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(!prevProps.open && this.props.open)) {
                  _context.next = 6;
                  break;
                }

                if (!(this.state.rect === null)) {
                  _context.next = 4;
                  break;
                }

                _context.next = 4;
                return this.setState({
                  rect: this.actionsBoxRef.current.getBoundingClientRect()
                });

              case 4:
                ml = this.props.messageListRect;

                if (this.props.mine) {
                  this.setState({
                    reverse: this.state.rect.left < ml.left ? true : false
                  });
                } else if (!this.props.mine) {
                  this.setState({
                    reverse: this.state.rect.right + 5 > ml.right ? true : false
                  });
                }

              case 6:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function componentDidUpdate(_x) {
        return _componentDidUpdate.apply(this, arguments);
      }

      return componentDidUpdate;
    }()
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          handleFlag = _this$props.handleFlag,
          handleMute = _this$props.handleMute,
          handleEdit = _this$props.handleEdit,
          handleDelete = _this$props.handleDelete,
          getMessageActions = _this$props.getMessageActions,
          t = _this$props.t;
      var messageActions = getMessageActions();
      return _react.default.createElement("div", {
        className: "str-chat__message-actions-box\n          ".concat(this.props.open ? 'str-chat__message-actions-box--open' : '', "\n          ").concat(this.props.mine ? 'str-chat__message-actions-box--mine' : '', "\n          ").concat(this.state.reverse ? 'str-chat__message-actions-box--reverse' : '', "\n        "),
        ref: this.actionsBoxRef
      }, _react.default.createElement("ul", {
        className: "str-chat__message-actions-list"
      }, messageActions.indexOf(_utils.MESSAGE_ACTIONS.flag) > -1 && _react.default.createElement("button", {
        onClick: handleFlag
      }, _react.default.createElement("li", {
        className: "str-chat__message-actions-list-item"
      }, t('Flag'))), messageActions.indexOf(_utils.MESSAGE_ACTIONS.mute) > -1 && _react.default.createElement("button", {
        onClick: handleMute
      }, _react.default.createElement("li", {
        className: "str-chat__message-actions-list-item"
      }, t('Mute'))), messageActions.indexOf(_utils.MESSAGE_ACTIONS.edit) > -1 && _react.default.createElement("button", {
        onClick: handleEdit
      }, _react.default.createElement("li", {
        className: "str-chat__message-actions-list-item"
      }, t('Edit Message'))), messageActions.indexOf(_utils.MESSAGE_ACTIONS.delete) > -1 && _react.default.createElement("button", {
        onClick: handleDelete
      }, _react.default.createElement("li", {
        className: "str-chat__message-actions-list-item"
      }, t('Delete')))));
    }
  }]);
  return MessageActionsBox;
}(_react.default.Component);

exports.MessageActionsBox = MessageActionsBox;
(0, _defineProperty2.default)(MessageActionsBox, "propTypes", {
  open: _propTypes.default.bool.isRequired,
  Message: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func, _propTypes.default.object]).isRequired,
  mine: _propTypes.default.bool,
  messageListRect: _propTypes.default.object,
  handleFlag: _propTypes.default.func,
  handleMute: _propTypes.default.func,
  handleEdit: _propTypes.default.func,
  handleDelete: _propTypes.default.func,
  getMessageActions: _propTypes.default.func
});
(0, _defineProperty2.default)(MessageActionsBox, "defaultProps", {
  open: false
});
exports.MessageActionsBox = MessageActionsBox = (0, _context2.withTranslationContext)(MessageActionsBox);
