// stream-chat-react/dist/Modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Modal = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Modal = function (_React$PureComponent) {
  (0, _inherits2.default)(Modal, _React$PureComponent);

  var _super = _createSuper(Modal);

  function Modal() {
    var _this;

    (0, _classCallCheck2.default)(this, Modal);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "innerRef", _react.default.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      if (!_this.innerRef.current.contains(e.target)) {
        _this.props.onClose();

        document.removeEventListener('keyPress', _this.handleEscKey, false);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Modal, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      document.addEventListener('keyPress', this.handleEscKey, false);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener('keyPress', this.handleEscKey, false);
    }
  }, {
    key: "handleEscKey",
    value: function handleEscKey(e) {
      if (e.keyCode === 27) {
        this.props.onClose();
        document.removeEventListener('keyPress', this.handleEscKey, false);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var openClasses = this.props.open ? 'str-chat__modal--open' : 'str-chat__modal--closed';
      return _react.default.createElement("div", {
        className: "str-chat__modal ".concat(openClasses),
        onClick: this.handleClick
      }, _react.default.createElement("div", {
        className: "str-chat__modal__close-button"
      }, "Close", _react.default.createElement("svg", {
        width: "10",
        height: "10",
        xmlns: "http://www.w3.org/2000/svg"
      }, _react.default.createElement("path", {
        d: "M9.916 1.027L8.973.084 5 4.058 1.027.084l-.943.943L4.058 5 .084 8.973l.943.943L5 5.942l3.973 3.974.943-.943L5.942 5z",
        fillRule: "evenodd"
      }))), _react.default.createElement("div", {
        className: "str-chat__modal__inner",
        ref: this.innerRef
      }, this.props.children));
    }
  }]);
  return Modal;
}(_react.default.PureComponent);

exports.Modal = Modal;
(0, _defineProperty2.default)(Modal, "propTypes", {
  onClose: _propTypes.default.func,
  open: _propTypes.default.bool
});
