// stream-chat-react/dist/ReactionSelector.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReactionSelector = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emojiMart = require("emoji-mart");

var _Avatar = require("./Avatar");

var _utils = require("../utils");

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ReactionSelector = function (_PureComponent) {
  (0, _inherits2.default)(ReactionSelector, _PureComponent);

  var _super = _createSuper(ReactionSelector);

  function ReactionSelector(props) {
    var _this;

    (0, _classCallCheck2.default)(this, ReactionSelector);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "showTooltip", function () {
      var _ref = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee2(e, users) {
        var target;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                target = e.target.getBoundingClientRect();
                _context2.next = 3;
                return _this.setState(function () {
                  return {
                    showTooltip: true,
                    users
                  };
                }, (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee() {
                  return _regenerator.default.wrap(function _callee$(_context) {
                    while (1) {
                      switch (_context.prev = _context.next) {
                        case 0:
                          _context.next = 2;
                          return _this.setTooltipPosition(target);

                        case 2:
                          _context.next = 4;
                          return _this.setArrowPosition(target);

                        case 4:
                        case "end":
                          return _context.stop();
                      }
                    }
                  }, _callee);
                })));

              case 3:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      return function (_x, _x2) {
        return _ref.apply(this, arguments);
      };
    }());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "hideTooltip", function () {
      _this.setState(function () {
        return {
          showTooltip: false,
          users: [],
          arrowPosition: 0,
          position: 0,
          positionCaculated: false
        };
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getUsersPerReaction", function (reactions, type) {
      return reactions && reactions.filter(function (item) {
        return item.type === type;
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getLatestUser", function (reactions, type) {
      var filtered = _this.getUsersPerReaction(reactions, type);

      if (filtered && filtered[0] && filtered[0].user) {
        return filtered[0].user;
      } else {
        return 'NotFound';
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getUserNames", function (reactions, type) {
      var filtered = _this.getUsersPerReaction(reactions, type);

      return filtered && filtered.map(function (item) {
        return item.user || 'NotFound';
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getContainerDimensions", function () {
      return _this.reactionSelector.current.getBoundingClientRect();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getToolTipDimensions", function () {
      return _this.reactionSelectorTooltip.current.getBoundingClientRect();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "setTooltipPosition", function () {
      var _ref3 = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee3(target) {
        var container, tooltip, position;
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return _this.getContainerDimensions();

              case 2:
                container = _context3.sent;
                _context3.next = 5;
                return _this.getToolTipDimensions();

              case 5:
                tooltip = _context3.sent;

                if (tooltip.width === container.width || tooltip.x < container.x) {
                  position = 0;
                } else {
                  position = target.left + target.width / 2 - container.left - tooltip.width / 2;
                }

                _context3.next = 9;
                return _this.setState(function () {
                  return {
                    position
                  };
                });

              case 9:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3);
      }));

      return function (_x3) {
        return _ref3.apply(this, arguments);
      };
    }());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "setArrowPosition", function () {
      var _ref4 = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee4(target) {
        var tooltip, position;
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                tooltip = _this.reactionSelectorTooltip.current.getBoundingClientRect();
                position = target.x - tooltip.x + target.width / 2;
                _context4.next = 4;
                return _this.setState(function () {
                  return {
                    arrowPosition: position,
                    positionCaculated: true
                  };
                });

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4);
      }));

      return function (_x4) {
        return _ref4.apply(this, arguments);
      };
    }());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderReactionItems", function () {
      var _this$props = _this.props,
          reaction_counts = _this$props.reaction_counts,
          latest_reactions = _this$props.latest_reactions;
      return _this.props.reactionOptions.map(function (reaction) {
        var users = _this.getUserNames(latest_reactions, reaction.id);

        var latestUser = _this.getLatestUser(latest_reactions, reaction.id);

        var count = reaction_counts && reaction_counts[reaction.id];
        return _react.default.createElement("li", {
          key: "item-".concat(reaction.id),
          className: "str-chat__message-reactions-list-item",
          "data-text": reaction.id,
          onClick: _this.props.handleReaction.bind((0, _assertThisInitialized2.default)(_this), reaction.id)
        }, Boolean(count) && _this.props.detailedView && _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
          className: "latest-user",
          onMouseEnter: function onMouseEnter(e) {
            return _this.showTooltip(e, users);
          },
          onMouseLeave: _this.hideTooltip
        }, latestUser !== 'NotFound' ? _react.default.createElement(_Avatar.Avatar, {
          image: latestUser.image,
          alt: latestUser.id,
          size: 20,
          name: latestUser.id
        }) : _react.default.createElement("div", {
          className: "latest-user-not-found"
        }))), _react.default.createElement(_emojiMart.NimbleEmoji, (0, _extends2.default)({
          emoji: reaction
        }, _utils.emojiSetDef, {
          data: _utils.emojiData
        })), Boolean(count) && _this.props.detailedView && _react.default.createElement("span", {
          className: "str-chat__message-reactions-list-item__count"
        }, count || ''));
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderUsers", function (users) {
      return users.map(function (user, i) {
        var text = user.name || user.id;

        if (i + 1 < users.length) {
          text += ', ';
        }

        return _react.default.createElement("span", {
          className: "latest-user-username",
          key: "key-".concat(i, "-").concat(user)
        }, text);
      });
    });
    _this.state = {
      showTooltip: false,
      users: [],
      position: 0,
      arrowPosition: 0,
      positionCalculated: false
    };
    _this.reactionSelector = _react.default.createRef();
    _this.reactionSelectorInner = _react.default.createRef();
    _this.reactionSelectorTooltip = _react.default.createRef();
    return _this;
  }

  (0, _createClass2.default)(ReactionSelector, [{
    key: "render",
    value: function render() {
      return _react.default.createElement("div", {
        className: "str-chat__reaction-selector ".concat(this.props.reverse ? 'str-chat__reaction-selector--reverse' : ''),
        ref: this.reactionSelector
      }, this.props.detailedView && _react.default.createElement("div", {
        className: "str-chat__reaction-selector-tooltip",
        ref: this.reactionSelectorTooltip,
        style: {
          visibility: this.state.showTooltip ? 'visible' : 'hidden',
          left: this.state.position,
          opacity: this.state.showTooltip && this.state.positionCaculated ? 1 : 0.01
        }
      }, _react.default.createElement("div", {
        className: "arrow",
        style: {
          left: this.state.arrowPosition
        }
      }), this.renderUsers(this.state.users)), _react.default.createElement("ul", {
        className: "str-chat__message-reactions-list"
      }, this.renderReactionItems()));
    }
  }]);
  return ReactionSelector;
}(_react.PureComponent);

exports.ReactionSelector = ReactionSelector;
(0, _defineProperty2.default)(ReactionSelector, "propTypes", {
  latest_reactions: _propTypes.default.array,
  reaction_counts: _propTypes.default.object,
  handleReaction: _propTypes.default.func.isRequired,
  detailedView: _propTypes.default.bool,
  reactionOptions: _propTypes.default.array,
  reverse: _propTypes.default.bool
});
(0, _defineProperty2.default)(ReactionSelector, "defaultProps", {
  detailedView: true,
  reactionOptions: _utils.defaultMinimalEmojis,
  reverse: false,
  emojiSetDef: _utils.emojiSetDef
});
