// stream-chat-react/dist/ReactionsList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReactionsList = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emojiMart = require("emoji-mart");

var _utils = require("../utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ReactionsList = function (_React$Component) {
  (0, _inherits2.default)(ReactionsList, _React$Component);

  var _super = _createSuper(ReactionsList);

  function ReactionsList(props) {
    var _this;

    (0, _classCallCheck2.default)(this, ReactionsList);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_renderReactions", function (reactions) {
      var reactionsByType = {};
      reactions.map(function (item) {
        if (reactions[item.type] === undefined) {
          return reactionsByType[item.type] = [item];
        } else {
          return reactionsByType[item.type] = [].concat((0, _toConsumableArray2.default)(reactionsByType[item.type]), [item]);
        }
      });

      var reactionsEmojis = _this.props.reactionOptions.reduce(function (acc, cur) {
        return _objectSpread({}, acc, {
          [cur.id]: cur
        });
      }, {});

      return Object.keys(reactionsByType).map(function (type) {
        return reactionsEmojis[type] ? _react.default.createElement("li", {
          key: reactionsEmojis[type].id
        }, _react.default.createElement(_emojiMart.NimbleEmoji, (0, _extends2.default)({
          emoji: reactionsEmojis[type]
        }, _utils.emojiSetDef, {
          size: 16,
          data: _utils.emojiData
        })), ' ', "\xA0") : null;
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_getReactionCount", function () {
      var reaction_counts = _this.props.reaction_counts;
      var count = null;

      if (reaction_counts !== null && reaction_counts !== undefined && Object.keys(reaction_counts).length > 0) {
        count = 0;
        Object.keys(reaction_counts).map(function (key) {
          return count += reaction_counts[key];
        });
      }

      return count;
    });
    return _this;
  }

  (0, _createClass2.default)(ReactionsList, [{
    key: "render",
    value: function render() {
      return _react.default.createElement("div", {
        className: "str-chat__reaction-list ".concat(this.props.reverse ? 'str-chat__reaction-list--reverse' : ''),
        onClick: this.props.onClick,
        ref: this.reactionList
      }, _react.default.createElement("ul", null, this._renderReactions(this.props.reactions), _react.default.createElement("li", null, _react.default.createElement("span", {
        className: "str-chat__reaction-list--counter"
      }, this._getReactionCount()))));
    }
  }]);
  return ReactionsList;
}(_react.default.Component);

exports.ReactionsList = ReactionsList;
(0, _defineProperty2.default)(ReactionsList, "propTypes", {
  reactions: _propTypes.default.array,
  reactionOptions: _propTypes.default.array,
  reverse: _propTypes.default.bool,
  reaction_counts: _propTypes.default.object
});
(0, _defineProperty2.default)(ReactionsList, "defaultProps", {
  reactionOptions: _utils.defaultMinimalEmojis,
  emojiSetDef: _utils.emojiSetDef,
  reverse: false
});
