// stream-chat-react/dist/SimpleReactionsList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleReactionsList = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emojiMart = require("emoji-mart");

var _utils = require("../utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var SimpleReactionsList = function (_React$PureComponent) {
  (0, _inherits2.default)(SimpleReactionsList, _React$PureComponent);

  var _super = _createSuper(SimpleReactionsList);

  function SimpleReactionsList() {
    var _this;

    (0, _classCallCheck2.default)(this, SimpleReactionsList);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      showTooltip: false,
      users: []
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "showTooltip", function () {
      _this.setState({
        showTooltip: true
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "hideTooltip", function () {
      _this.setState({
        showTooltip: false
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleReaction", function (e, type) {
      if (e !== undefined && e.preventDefault) {
        e.preventDefault();
      }

      _this.props.handleReaction(type);

      _this.setUsernames(type);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getReactionCount", function () {
      var reaction_counts = _this.props.reaction_counts;
      var count = null;

      if (reaction_counts !== null && reaction_counts !== undefined && Object.keys(reaction_counts).length > 0) {
        count = 0;
        Object.keys(reaction_counts).map(function (key) {
          return count += reaction_counts[key];
        });
      }

      return count;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderUsers", function (users) {
      return users.map(function (user, i) {
        var text = user;

        if (i + 1 < users.length) {
          text += ', ';
        }

        return _react.default.createElement("span", {
          className: "latest-user-username",
          key: "key-".concat(i, "-").concat(user)
        }, text);
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getReactionsByType", function (reactions) {
      var reactionsByType = {};
      reactions.map(function (item) {
        if (reactionsByType[item.type] === undefined) {
          return reactionsByType[item.type] = [item];
        } else {
          return reactionsByType[item.type] = [].concat((0, _toConsumableArray2.default)(reactionsByType[item.type]), [item]);
        }
      });
      return reactionsByType;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderReactions", function (reactions) {
      var reactionsByType = _this.getReactionsByType(reactions);

      var reactionsEmojis = _this.props.reactionOptions.reduce(function (acc, cur) {
        return _objectSpread({}, acc, {
          [cur.id]: cur
        });
      }, {});

      return Object.keys(reactionsByType).map(function (type, i) {
        return _react.default.createElement("li", {
          className: "str-chat__simple-reactions-list-item",
          key: "".concat(reactionsByType[type][0].id, "-").concat(i),
          onClick: function onClick(e) {
            return _this.handleReaction(e, type);
          }
        }, _react.default.createElement("span", {
          onMouseEnter: function onMouseEnter() {
            return _this.setUsernames(type);
          }
        }, _react.default.createElement(_emojiMart.NimbleEmoji, (0, _extends2.default)({
          emoji: reactionsEmojis[type]
        }, _utils.emojiSetDef, {
          size: 13,
          data: _utils.emojiData
        })), "\xA0"));
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getUsernames", function (reactions) {
      return reactions.map(function (item) {
        return item.user !== null ? item.user.name || item.user.id : 'null';
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "setUsernames", function (type) {
      var reactionsByType = _this.getReactionsByType(_this.props.reactions);

      var reactions = reactionsByType[type];

      var users = _this.getUsernames(reactions);

      _this.setState({
        users
      }, function () {
        return _this.showTooltip();
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderUsernames", function (users) {
      return users.join(', ');
    });
    return _this;
  }

  (0, _createClass2.default)(SimpleReactionsList, [{
    key: "render",
    value: function render() {
      var reactions = this.props.reactions;

      if (!reactions || reactions.length === 0) {
        return null;
      }

      return _react.default.createElement("ul", {
        className: "str-chat__simple-reactions-list",
        onMouseLeave: this.hideTooltip
      }, this.state.showTooltip && _react.default.createElement("div", {
        className: "str-chat__simple-reactions-list-tooltip",
        ref: this.reactionSelectorTooltip
      }, _react.default.createElement("div", {
        className: "arrow"
      }), this.renderUsernames(this.state.users)), this.renderReactions(reactions), reactions.length !== 0 && _react.default.createElement("li", {
        className: "str-chat__simple-reactions-list-item--last-number"
      }, this.getReactionCount()));
    }
  }]);
  return SimpleReactionsList;
}(_react.default.PureComponent);

exports.SimpleReactionsList = SimpleReactionsList;
(0, _defineProperty2.default)(SimpleReactionsList, "propTypes", {
  reactions: _propTypes.default.array,
  reaction_counts: _propTypes.default.object,
  showTooltip: _propTypes.default.bool,
  reactionOptions: _propTypes.default.array,
  handleReaction: _propTypes.default.func
});
(0, _defineProperty2.default)(SimpleReactionsList, "defaultProps", {
  showTooltip: true,
  reactionOptions: _utils.defaultMinimalEmojis,
  emojiSetDef: _utils.emojiSetDef
});
